providers/implementations/ciphers/cipher_rc2.c

```diff
@@ -140,13 +144,23 @@ static int rc2_get_ctx_params(void *vctx, OSSL_PARAM params[])
             ERR_raise(ERR_LIB_PROV, ERR_R_ASN1_LIB);
             return 0;
         }
+
         /*
          * IF the caller has a buffer, we pray to the gods they got the
          * size right.  There's no way to tell the i2d functions...
          */
-        i = i2d_ASN1_TYPE(type, dd);
-        if (i >= 0)
-            p->return_size = (size_t)i;
+        i = i2d_ASN1_TYPE(type, dd1);
+        if (p1 != NULL && i >= 0)
+            p1->return_size = (size_t)i;
+
+        /*
+         * If the buffers differ, redo the i2d on the second buffer.
+         * Otherwise, just use |i| as computed above
+         */
+        if (d1 != d2)
+            i = i2d_ASN1_TYPE(type, dd2);
```

Comment:
Ah right, it needs to be conditioned on the presence of the corresponding parameter (i.e. *p2*)...  and even the lines above, with the production of `dd1`, should be conditioned on *p1* != NULL

Reply:
Oh wait...  actually, I think you are incorrect here, 'cause when *d1* is NULL, *dd1* is also NULL, and `i2d_ASN1_TYPE()` will then only calculate the size, which is exactly what we want.

Reply:
... so worst case scenario, *i* is calculated twice, and buffers are never allocated, 'cause *dd1* will only be non-NULL when *d1* is non-NULL, and the same for *dd2* and *d2*, and that's when anything get written to them.

In other words, the caller allocates the buffers, not this code.

Reply:
ok, Admittedly, I was having trouble tracing through all the macro definitions to find where i2d_ASN1_TYPE was defined.  The docs [here](https://docs.openssl.org/master/man3/d2i_X509/#description) indicate that i2d_TYPE allocates storage if the 2nd argument points to a NULL pointer.  Is that incorrect?  From the docs:
```
i2d_TYPE() encodes the structure pointed to by a into DER format. If ppout is not NULL, it writes the DER encoded data to the buffer at *ppout, and increments it to point after the data just written. If the return value is negative an error occurred, otherwise it returns the length of the encoded data.

If *ppout is NULL memory will be allocated for a buffer and the encoded data written to it. In this case *ppout is not incremented and it points to the start of the data just written.
```

Reply:
ok, so I get it now.  I found it here I think.  i2d_TYPE (nee i2d_ASN1_TYPE gets declared via IMPLEMENT_ASN1_FUNCTIONS_fname

That defines i2d_ASN1_TYPE as a call to ASN1_item_i2d->asn1_item_flags_i2d, which allocates a buffer for out if *out is NULL

Back in rc2_get_ctx_params we do this:

```
unsigned char **dd2 = d2 == NULL ? NULL : &d2;
```
Which by my read means dd2 will either be NULL or the address of d2

later when we make this call
```
if (d1 != d2)
            i = i2d_ASN1_TYPE(type, dd2);
```
since dd2 May either be NULL itself or point to a NULL pointer, we skip the allocation in asn1_item_flags_i2d.  Makes sense, you're right.

Reply:
> ... or point to a NULL pointer

Well, no.  *dd2* either *is NULL*, or points to a *non-NULL* pointer.

Reply:
> ok, Admittedly, I was having trouble tracing through all the macro definitions to find where i2d_ASN1_TYPE was defined. The docs [here](https://docs.openssl.org/master/man3/d2i_X509/#description) indicate that i2d_TYPE allocates storage if the 2nd argument points to a NULL pointer. Is that incorrect? From the docs:
> 
> ```
> i2d_TYPE() encodes the structure pointed to by a into DER format. If ppout is not NULL, it writes the DER encoded data to the buffer at *ppout, and increments it to point after the data just written. If the return value is negative an error occurred, otherwise it returns the length of the encoded data.
> 
> If *ppout is NULL memory will be allocated for a buffer and the encoded data written to it. In this case *ppout is not incremented and it points to the start of the data just written.
> ```

Do note very precisely how it's written.  There's a difference between *ppout* and *\*ppout*.  So you basically have three states to keep track of:

- *ppout* == NULL : no encoding happens, but the byte count of the encoding is still calculated.
- *\*ppout* == NULL : a buffer is allocated and assigned to *\*ppout*, encoding happens
- *\*ppout* != NULL : encoding happens (trusting the caller to have allocated enough space)